pymc/distributions/distribution.py

```diff
@@ -861,3 +865,77 @@ def default_moment(rv, size, *rv_inputs, rv_name=None, has_fallback=False, ndim_
             f"Please provide a moment function when instantiating the {rv_name} "
             "random variable."
         )
+
+
+def change_dist_size(
+    dist: TensorVariable,
+    new_size: PotentialShapeType,
+    expand: Optional[bool] = False,
+) -> TensorVariable:
+    """Change or expand the size of a Distribution.
+
+    Parameters
+    ==========
+    dist
+        The old Distribution output.
+    new_size
+        The new size.
+    expand:
+        Expand the existing size by `new_size`.
+
+    """
+    # Check the dimensionality of the `new_size` kwarg
+    new_size_ndim = np.ndim(new_size)
+    if new_size_ndim > 1:
+        raise ShapeError("The `new_size` must be ≤1-dimensional.", actual=new_size_ndim)
+    elif new_size_ndim == 0:
+        new_size = (new_size,)
+
+    new_dist = _change_dist_size(dist.owner.op, dist, new_size=new_size, expand=expand)
+
+    new_dist.name = dist.name
+    for k, v in dist.tag.__dict__.items():
+        new_dist.tag.__dict__.setdefault(k, v)
+
+    if config.compute_test_value != "off":
+        compute_test_value(new_dist)
+
+    return new_dist
+
+
+@singledispatch
+def _change_dist_size(op, dist, new_size, expand):
+    raise NotImplementedError(
+        f"Variable {dist} of type {op} has no _change_dist_size implementation."
+    )
+
+
+@_change_dist_size.register(RandomVariable)
+def change_rv_size(op, rv, new_size, expand):
+    # Extract the RV node that is to be resized
+    rv_node = rv.owner
+    rng, size, dtype, *dist_params = rv_node.inputs
+
+    if expand:
+        shape = tuple(rv_node.op._infer_shape(size, dist_params))
+        size = shape[: len(shape) - rv_node.op.ndim_supp]
+        new_size = tuple(new_size) + tuple(size)
+
+    # Make sure the new size is a tensor. This dtype-aware conversion helps
+    # to not unnecessarily pick up a `Cast` in some cases (see #4652).
+    new_size = at.as_tensor(new_size, ndim=1, dtype="int64")
+
+    new_rv_node = rv_node.op.make_node(rng, new_size, dtype, *dist_params)
+    new_rv = new_rv_node.outputs[-1]
+
+    # Update "traditional" rng default_update, if that was set for old RV
+    default_update = getattr(rng, "default_update", None)
+    if default_update is not None and default_update is rv_node.outputs[0]:
+        rng.default_update = new_rv_node.outputs[0]
+
+    return new_rv
+
+
+@_change_dist_size.register(SpecifyShape)
```

Comment:
Should this also dispatch asserts and prints?

Reply:
I never used prints so I don't know. I am not sure about asserts, what if they are related to shapes of inputs/outputs. We can't easily know what they are about. If a variable requires Asserts to work properly, I would suggest they are added in the inner graph of a SymbolicRV. The custom `_change_dist_size` dispatch function will then know what asserts are needed in the recreated variable.